lib/Doctrine/DBAL/Query/BulkInsertQuery.php

```diff
@@ -0,0 +1,271 @@
+     *                                   (named and indexed).
+     *
+     * @todo add support for expressions.
+     */
+    public function addValues(array $values, array $types = array())
+    {
+        $valueSet = array();
+
+        if (empty($this->columns)) {
+            foreach ($values as $index => $value) {
+                $this->parameters[] = $value; // todo: allow expressions.
+                $this->types[] = isset($types[$index]) ? $types[$index] : null;
+                $valueSet[] = '?'; // todo: allow expressions.
+            }
+
+            $this->values[] = $valueSet;
+
+            return;
+        }
+
+        foreach ($this->columns as $index => $column) {
+            $namedValue = isset($values[$column]) || array_key_exists($column, $values);
+            $positionalValue = isset($values[$index]) || array_key_exists($index, $values);
+
+            if ( ! $namedValue && ! $positionalValue) {
+                throw new \InvalidArgumentException(
+                    sprintf('No value specified for column %s (index %d).', $column, $index)
+                );
+            }
+
+            if ($namedValue && $positionalValue && $values[$column] !== $values[$index]) {
+                throw new \InvalidArgumentException(
+                    sprintf('Multiple values specified for column %s (index %d).', $column, $index)
+                );
+            }
+
+            $this->parameters[] = $namedValue ? $values[$column] : $values[$index]; // todo: allow expressions.
+            $valueSet[] = '?'; // todo: allow expressions.
+
+            $namedType = isset($types[$column]);
+            $positionalType = isset($types[$index]);
+
+            if ($namedType && $positionalType && $types[$column] !== $types[$index]) {
+                throw new \InvalidArgumentException(
+                    sprintf('Multiple types specified for column %s (index %d).', $column, $index)
+                );
+            }
+
+            if ($namedType) {
+                $this->types[] = $types[$column];
+
+                continue;
+            }
+
+            if ($positionalType) {
+                $this->types[] = $types[$index];
+
+                continue;
+            }
+
+            $this->types[] = null;
+        }
+
+        $this->values[] = $valueSet;
+    }
+
+    /**
+     * Executes this INSERT query using the bound parameters and their types.
+     *
+     * @return integer The number of affected rows.
+     *
+     * @throws \LogicException if this query contains more rows than acceptable
+     *                         for a single INSERT statement by the underlying platform.
+     */
+    public function execute()
+    {
+        $platform = $this->connection->getDatabasePlatform();
+        $insertMaxRows = $platform->getInsertMaxRows();
+
+        if ($insertMaxRows > 0 && count($this->values) > $insertMaxRows) {
+            throw new \LogicException(
```

Comment:
I think a better way to handle this condition in `addValues` method and throw some kind of specific exception to allow clients write code in a way:
```
try {
  $bulk->addValues(...);
} catch (MaxValuesReachedException $e) {
  $bulk->execute();
}
```